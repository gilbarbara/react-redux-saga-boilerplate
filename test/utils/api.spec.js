import fetchMock from 'fetch-mock';
import { request } from 'utils/api';

describe('api', () => {
  afterEach(() => {
    fetchMock.restore();
  });

  it('should fail without param', () => {
    expect(() => {
      request();
    }).toThrow('Error! You must pass `endpoint`');
  });

  it('should fail to a POST without payload', () => {
    expect(() => {
      request({ endpoint: 'http://example.com/token', method: 'POST' });
    }).toThrow('Error! You must pass `payload`');
  });

  it('should execute a GET successfully with json', done => {
    fetchMock.mock('http://example.com/token', {
      status: 200,
      body: { hello: 'world' },
      headers: { 'Content-Type': 'application/json' },
    });

    request({ endpoint: 'http://example.com/token' })
      .then(data => {
        expect(data).toMatchSnapshot();
        done();
      });
  });

  it('should reject for a server error with JSON response', done => {
    fetchMock.mock('http://example.com/token', {
      body: { error: 'FAILED' },
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });

    request({ endpoint: 'http://example.com/token' })
      .catch(error => {
        expect(error).toMatchSnapshot();
        done();
      });
  });

  it('should reject for a server error with no response', done => {
    fetchMock.mock('http://example.com/token', {
      status: 500,
    });

    request({ endpoint: 'http://example.com/token' })
      .catch(error => {
        expect(error).toMatchSnapshot();
        done();
      });
  });

  it('should reject for a not found error', done => {
    fetchMock.mock('http://example.com/token', 404, {
      error: 'FAILED',
      headers: { 'Content-Type': 'application/json' },
    });

    request({ endpoint: 'http://example.com/token' })
      .catch(error => {
        expect(error).toMatchSnapshot();
        done();
      });
  });

  it('should execute a POST successfully', done => {
    fetchMock.post('http://example.com/token', {
      body: 'ok',
      status: 201,
    });

    request({ endpoint: 'http://example.com/token', method: 'POST', payload: { a: 1 } })
      .then(data => {
        expect(data).toMatchSnapshot();
        done();
      });
  });
});
