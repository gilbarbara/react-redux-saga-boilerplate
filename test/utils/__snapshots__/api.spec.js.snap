// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api should execute a GET successfully with json 1`] = `
Object {
  "hello": "world",
}
`;

exports[`api should execute a POST successfully 1`] = `"ok"`;

exports[`api should reject for a not found error 1`] = `
Object {
  "code": 404,
  "data": "",
  "error": "Not Found",
  "status": "FAIL",
}
`;

exports[`api should reject for a server error with JSON response 1`] = `
Object {
  "code": 500,
  "data": Object {
    "error": "FAILED",
  },
  "error": "Internal Server Error",
  "status": "FAIL",
}
`;

exports[`api should reject for a server error with no response 1`] = `
Object {
  "code": 500,
  "data": "",
  "error": "Internal Server Error",
  "status": "FAIL",
}
`;
